src/librustc/ty/sty.rs

```diff
@@ -2086,15 +2087,66 @@ impl<'tcx> LazyConst<'tcx> {
 }
 
 /// Typed constant value.
-#[derive(Copy, Clone, Debug, Hash, RustcEncodable, RustcDecodable, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Copy, Clone, Debug, RustcEncodable, RustcDecodable, Eq, Ord, PartialOrd)]
 pub struct Const<'tcx> {
     pub ty: Ty<'tcx>,
 
-    pub val: ConstValue<'tcx>,
+    /// This field is an optimization for caching commonly needed values of constants like `usize`
+    /// (or other integers for enum discriminants) and slices (e.g. from `b"foo"` and `"foo"`
+    /// literals)
+    pub val: ConstValue,
+
+    /// The actual backing storage of the constant and a pointer which can be resolved back to the
+    /// `allocation` field
+    ///
+    /// Can be `None` for trivial constants created from literals or directly. Is always `Some` for
+    /// aggregate constants or any named constant that you can actually end up taking a reference
+    /// to. This will get unwrapped in situations where we do know that it's a referencable
+    pub alloc: Option<(&'tcx Allocation, Pointer)>,
```

Comment:
Hm, I think I'd prefer if this would not have to be an `Option` -- but if that has a non-zero cost, and if this does not actually cause any problems in practice, then whatever.

Reply:
It's actually somewhat problematic to not use an option, because we generate a bunch of simple constants like 1 and 0 in the match code. These constants never leak to the rest of the compiler afaik, but it simplifies the code a lot to use the same datastructure

Reply:
Okay, makes sense, I guess.